### PR TITLE
Support warm-start in tfra.dynamic_embedding

### PR DESCRIPTION
**Pull request by dakabang**
# Description

In tensorflow we use tf.train.Checkpoint(tf2) or tf.compat.v1.train.Saver(tf1) for saving/restoring model, but I found it a little hard when I use these APIs to restore part of the model in tfra to perform transfer learning. In recommendation system, it's a common trick to speed up model convergence using embedding warmup, and the dense part of the model is trained from scratch, hence partial restore. The demonstrate the API design for partial restore are as follows.

```python
restore_op = tfra.dynamic_embedding.warm_start(ckpt_to_initialize_from='/path/to/ckpt/', vars_to_warm_start='.*embeddings.*')
with tf.compat.v1.Session() as sess:
   sess.run(restore_op)
   ...
```

# Changes

- Add tfra.dynamic_embedding.warm_start to support warm-start in tfra model.

### CODE REVIEW

**Comment by rhdong** [comment]
> 

Hi @dakabang , please sign the CLA by following the guide, thank you!
❌ https://github.com/tensorflow/recommenders-addons/commit/18f7c4f9c5a5b7f8b9d503ce00729e81feb3b5dd PR Opener: @dakabang <zha*****ng​@yeah.net>
❌ https://github.com/tensorflow/recommenders-addons/commit/18f7c4f9c5a5b7f8b9d503ce00729e81feb3b5dd Author: <eth******han​@vipshop.com>

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/warm_start_util.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
+# Copyright 2020 The TensorFlow Recommenders-Addons Authors.
```
Comment (rhdong) [inline tensorflow_recommenders_addons/dynamic_embedding/python/ops/warm_start_util.py:1]: 2020->2022

File: tensorflow_recommenders_addons/dynamic_embedding/__init__.py [hunk 2]
```diff
@@ -80,6 +80,8 @@
     Variable,)
 from tensorflow_recommenders_addons.dynamic_embedding.python.ops.dynamic_embedding_variable import (
     GraphKeys,)
+from tensorflow_recommenders_addons.dynamic_embedding.python.ops.warm_start_util import (
```
Comment (rhdong) [inline tensorflow_recommenders_addons/dynamic_embedding/__init__.py:83]: Please squash two commits to one.

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/warm_start_util.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+from tensorflow.python.framework import errors
+from tensorflow.python.framework import ops
+from tensorflow.python.ops import state_ops
+from tensorflow.python.ops import variable_scope
+from tensorflow.python.ops import variables as variables_lib
+from tensorflow.python.platform import tf_logging as logging
+from tensorflow.python.training import checkpoint_ops
+from tensorflow.python.training import checkpoint_utils
+from tensorflow.python.training import saver as saver_lib
+from tensorflow.python.training.saving import saveable_object_util
```
Comment (rhdong) [inline tensorflow_recommenders_addons/dynamic_embedding/python/ops/warm_start_util.py:31]: Please sort alphabetically

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/warm_start_util.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+  de_variables = []
+  for v in list_of_vars:
+    t = [v] if not isinstance(v, list) else v
+    de_variables.append(v)
+
+  return de_variables
+
+
+def warm_start(ckpt_to_initialize_from,
```
Comment (rhdong) [inline tensorflow_recommenders_addons/dynamic_embedding/python/ops/warm_start_util.py:69]: Please add function comments by Tensorflow style, You can refer to the TF warm start API.

File: tensorflow_recommenders_addons/dynamic_embedding/python/kernel_tests/warm_start_util_test.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+from tensorflow.python.training import device_setter
+from tensorflow.python.training import server_lib
+from tensorflow.python.util import compat
+from tensorflow.python.training import saver
+from tensorflow.python.ops import resources
+
+import tensorflow as tf
+
+
+class WarmStartHook(tf.estimator.SessionRunHook):
```
Comment (rhdong) [inline tensorflow_recommenders_addons/dynamic_embedding/python/kernel_tests/warm_start_util_test.py:64]: Why not make it a user API?

File: tensorflow_recommenders_addons/dynamic_embedding/python/kernel_tests/warm_start_util_test.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+  def after_create_session(self, session, coord):
+    session.run(self._restore_op)
+
+
+@test_util.deprecated_graph_mode_only
+class WarmStartUtilTest(test.TestCase):
+
+  def _test_warm_start(self, num_shards, use_regex):
+    devices = ["/cpu:0" for _ in range(num_shards)]
```
Comment (rhdong) [inline tensorflow_recommenders_addons/dynamic_embedding/python/kernel_tests/warm_start_util_test.py:83]: Please add 'GPU' support by referring [here](https://github.com/tensorflow/recommenders-addons/blob/master/tensorflow_recommenders_addons/dynamic_embedding/python/kernel_tests/dynamic_embedding_optimizer_test.py#L71)

File: tensorflow_recommenders_addons/dynamic_embedding/python/kernel_tests/warm_start_util_test.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+                                   devices=devices,
+                                   initializer=0.0)
+      ids = constant_op.constant(id_list, dtype=dtypes.int64)
+      vals = constant_op.constant(val_list, dtype=dtypes.float32)
+      self.evaluate(embeddings.upsert(ids, vals))
+      save = saver.Saver(var_list=[embeddings])
+      save.save(sess, ckpt_prefix)
+
+    with self.session(graph=ops.Graph()) as sess:
+      embeddings = de.get_variable("embeddings",
```
Comment (rhdong) [inline tensorflow_recommenders_addons/dynamic_embedding/python/kernel_tests/warm_start_util_test.py:101]: recommend using different name for testing different name function in one case, or you'd better to write another case.

File: tensorflow_recommenders_addons/dynamic_embedding/python/kernel_tests/warm_start_util_test.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+
+    predictor = tf.estimator.Estimator(model_fn=_model_fn)
+    predictions = predictor.predict(_input_fn)
+    pred_vals = []
+    for pred in predictions:
+      pred_vals.append(pred)
+    self.assertAllEqual(pred_vals, val_list)
+
+  def test_warm_start(self):
+    for num_shards in [1, 3, 10, 100]:
```
Comment (rhdong) [inline tensorflow_recommenders_addons/dynamic_embedding/python/kernel_tests/warm_start_util_test.py:172]: I believe too many shards are unnecessary to be tested. `[1,3]` is enough.

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/warm_start_util.py (hunk 4 above)
Comment (rhdong) [inline tensorflow_recommenders_addons/dynamic_embedding/python/ops/warm_start_util.py:69]: I don't know how is the A regular expression (string) supported here? Could you clarify for me, thank so much~

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/warm_start_util.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+    saveables = saveable_object_util.validate_and_slice_inputs([var])
+    for saveable in saveables:
+      restore_specs = []
+      for spec in saveable.specs:
+        restore_specs.append((spec.name, spec.slice_spec, spec.dtype))
+
+      names, slices, dtypes = zip(*restore_specs)
+      # Load all tensors onto CPU 0 for compatibility with existing code.
+      with ops.colocate_with(saveable.op._resource_handle.op):
```
Comment (rhdong) [inline tensorflow_recommenders_addons/dynamic_embedding/python/ops/warm_start_util.py:103]: Maybe it causes the memory in worker OOM?

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/warm_start_util.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+  for variable in de_variables:
+    var_name = variable.name
+    prev_var_name = var_name_to_prev_var_name.get(var_name)
+    if prev_var_name:
+      logging.debug('Warm-starting variable: {}; prev_var_name: {}'.format(
+          var_name, prev_var_name or "Unchanged"))
+    else:
+      prev_var_name = var_name
+
+    store_vars = variable_scope._get_default_variable_store()._vars
```
Comment (ccsquare) [inline tensorflow_recommenders_addons/dynamic_embedding/python/ops/warm_start_util.py:89]: It seems that "store_vars" here only stores variables with current var names. So store_vars.get(prev_var_name, None) will always return None when prev_var_name and var name are different?

File: tensorflow_recommenders_addons/dynamic_embedding/python/kernel_tests/warm_start_util_test.py [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+    self._restore_op = de.warm_start(
+        ckpt_to_initialize_from=self._ckpt_to_initialize_from,
+        vars_to_warm_start=self._vars_to_warm_start)
+
+  def after_create_session(self, session, coord):
+    session.run(self._restore_op)
+
+
+@test_util.deprecated_graph_mode_only
+class WarmStartUtilTest(test.TestCase):
```
Comment (ccsquare) [inline tensorflow_recommenders_addons/dynamic_embedding/python/kernel_tests/warm_start_util_test.py:80]: Maybe some tests with "var_name_to_prev_var_name" is set can be provided?

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/warm_start_util.py (hunk 4 above)
Comment (dakabang) [reply tensorflow_recommenders_addons/dynamic_embedding/python/ops/warm_start_util.py:69]: In here I use ops.get_collection's scope argument to support regular expression match, the resulting list of ops.get_collection is filtered to include only items whose name attribute matches scope using re.match

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/warm_start_util.py (hunk 9 above)
Comment (dakabang) [reply tensorflow_recommenders_addons/dynamic_embedding/python/ops/warm_start_util.py:140]: The comment isn't right, I place restore op on the same device as cuckoo hashtable op which is placed on PS in distributed training, this will avoid OOM in worker.

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/warm_start_util.py (hunk 4 above)
Comment (rhdong) [reply tensorflow_recommenders_addons/dynamic_embedding/python/ops/warm_start_util.py:69]: Thank you for your clarify!

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/warm_start_util.py [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+
+
+class WarmStartHook(SessionRunHook):
+  """Warm-start hook for tf.estimator.Estimator
+    """
+
+  def __init__(self,
+               ckpt_to_initialize_from,
+               vars_to_warm_start,
+               var_name_to_prev_var_name=None):
```
Comment (rhdong) [inline tensorflow_recommenders_addons/dynamic_embedding/python/ops/warm_start_util.py:154]: Please add function comments.
